edge-modules/DirectMethodSender/Program.cs

```diff
@@ -21,32 +21,58 @@ public static async Task<int> MainAsync()
         {
             Logger.LogInformation($"Starting DirectMethodSender with the following settings:\r\n{Settings.Current}");
 
-            ModuleClient moduleClient = null;
+            (CancellationTokenSource cts, ManualResetEventSlim completed, Option<object> handler) = ShutdownHandler.Init(TimeSpan.FromSeconds(5), Logger);
             try
             {
-                (CancellationTokenSource cts, ManualResetEventSlim completed, Option<object> handler) = ShutdownHandler.Init(TimeSpan.FromSeconds(5), Logger);
+                DirectMethodSenderBase directMethodClient;
+                switch (Settings.Current.InvocationSource)
+                {
+                    case InvocationSource.Local:
+                        // Implicit OpenAsync()
+                        directMethodClient = await DirectMethodLocalSender.CreateAsync(
+                                Settings.Current.TransportType,
+                                ModuleUtil.DefaultTimeoutErrorDetectionStrategy,
+                                ModuleUtil.DefaultTransientRetryStrategy,
+                                Logger);
+                        break;
+
+                    case InvocationSource.Cloud:
+                        // Implicit OpenAsync()
+                        directMethodClient = await DirectMethodCloudSender.CreateAsync(
+                                Settings.Current.ServiceClientConnectionString.Expect(() => new ArgumentException("ServiceClientConnectionString is null")),
+                                (Microsoft.Azure.Devices.TransportType)Settings.Current.TransportType,
+                                Logger);
+                        break;
+
+                    default:
+                        throw new NotImplementedException("Invalid InvocationSource type");
+                }
 
-                moduleClient = await ModuleUtil.CreateModuleClientAsync(
+                Option<Uri> analyzerUrl = Settings.Current.AnalyzerUrl;
+                ModuleClient reportClient = await ModuleUtil.CreateModuleClientAsync(
                     Settings.Current.TransportType,
                     ModuleUtil.DefaultTimeoutErrorDetectionStrategy,
                     ModuleUtil.DefaultTransientRetryStrategy,
                     Logger);
+                while (!cts.Token.IsCancellationRequested)
```

Comment:
I think you need to have try-catch inside this while loop; as for direct method cloud sender in long haul we would like to continue making direct method call even there is an unexpected exception happened.

Reply:
The `await directMethodClient.InvokeDirectMethodAsync(cts);` already has the internal try-catch to handle the scenario